google-datacatalog-qlik-connector/src/google/datacatalog_connectors/qlik/scrape/base_engine_api_helper.py

```diff
@@ -0,0 +1,146 @@
+#!/usr/bin/python
+#
+# Copyright 2020 Google LLC
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#      http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import abc
+import asyncio
+import json
+import logging
+import threading
+
+from urllib.parse import urlparse
+import websockets
+
+from google.datacatalog_connectors.qlik.scrape import constants
+
+
+class BaseEngineAPIHelper(abc.ABC):
+    """The base class for all Engine API Helpers."""
+
+    def __init__(self, server_address, auth_cookie):
+        # The server address starts with an http/https scheme. The below
+        # statement replaces the original scheme with 'wss', which is used for
+        # secure websockets communication.
+        self.__base_api_endpoint = f'wss://{urlparse(server_address).hostname}'
+        self.__auth_cookie = auth_cookie
+        self.__common_headers = {
+            constants.XRFKEY_HEADER_NAME: constants.XRFKEY,
+        }
+        self.__requests_counter = 0
+
+    def _connect_websocket(self, app_id):
+        """Opens a websocket connection.
+
+        Args:
+            app_id:
+              An App ID to be appended to the Engine API URLs when making
+              websocket connections and then get an isolated Qlik Engine
+              session for each Qlik App (see [Connecting to the Qlik Engine
+              JSON API](https://help.qlik.com/en-US/sense-developer/September2020/Subsystems/EngineAPI/Content/Sense_EngineAPI/GettingStarted/connecting-to-engine-api.htm)).  # noqa E510
+
+        Returns:
+            An awaiting function that yields a :class:`WebSocketClientProtocol`
+            which can then be used to send and receive messages.
+        """
+        uri = f'{self.__base_api_endpoint}/app/{app_id}' \
+              f'?Xrfkey={constants.XRFKEY}'
+
+        auth_cookie = self.__auth_cookie
+        headers = self.__common_headers.copy()
+        # Format the header value as <key>=<value> string.
+        headers['Cookie'] = f'{auth_cookie.name}={auth_cookie.value}'
+
+        return websockets.connect(uri=uri, extra_headers=headers)
+
+    def _generate_request_id(self):
```

Comment:
Exploring the Qlik Sense UI, I've noticed they use sequential values. They also use them in the Engine API Explorer. It's simples and easier to understand the whole communication session.

Reply:
This does not seem to be thread safe, since it uses a instance(self) variable to store the counter state.

Assuming that the `base_engine_api_helper` will be called in parallel... Unless each thread uses its own instance of the `base_engine_api_helper.py` this might generate some concurrency issues.

Reply:
Makes sense. I'm adding a `threading.Lock()` to control concurrent access.